src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,482 @@
+                   -- rename tabledata to rowdata or just data
+  versionedTablesInsert name b
+
+type ParentHash = BlockHash
+
+handleVersion :: BlockHeight -> Maybe ParentHash -> BlockHandler SQLiteEnv ()
+handleVersion _ Nothing = return ()
+handleVersion bRestore (Just hsh) = do
+  bCurrent <- do
+    r <- callDb $ \ db -> qry_ db "SELECT max(blockheight) AS current_block_height FROM BlockHistory;" [RInt]
+    SInt bh <- liftIO $ expectSingleRowCol "detectVersionChange (block):" r
+    return $ BlockHeight (fromIntegral bh)
+  vCurrent <- do
+    r <- callDb $ \db -> qry_ db "SELECT max(version) AS current_version FROM VersionHistory;" [RInt]
+    SInt version <- liftIO $ expectSingleRowCol "detectVersionChange (version):" r
+    return $ ReorgVersion version
+
+  -- REWRITE ERROR WITH INTERNALERROR PREFIX
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <- callDb $ \db -> do
+        res <- qry db "SELECT COUNT(*)\
+           \ FROM BlockHistory\
+           \ WHERE blockheight = ?\
+           \ AND hash = ?;"
+            [SInt $ fromIntegral $ pred bRestore, SBlob (Data.Serialize.encode hsh)] [RInt]
+            >>= expectSingleRowCol "detectVersionChange (historyInvariant):"
+        return res
+
+  when (historyInvariant /= SInt 1) $
+    internalError "History invariant violation"
+
+  case compare bRestore (bCurrent + 1) of
+   GT -> internalError "PactInternalError: detectVersionChange: Block_Restore invariant violation!"
+   EQ -> assign bsBlockVersion (BlockVersion bRestore vCurrent)
+   LT -> do
+     let bvEnv = (BlockVersion bRestore (succ vCurrent))
+     assign bsBlockVersion bvEnv
+     tableMaintenanceRowsVersionedTables bvEnv
+     dropVersionedTables bvEnv
+     deleteHistory bvEnv
+
+tableMaintenanceRowsVersionedTables :: BlockVersion -> BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedTables bv = do
+  tableMaintenanceRowsVersionedSystemTables bv
+  tableMaintenanceRowsVersionedUserTables bv
+
+tableMaintenanceRowsVersionedSystemTables :: BlockVersion -> BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedSystemTables (BlockVersion bh _) = callDb $ \db -> do
+  exec' db "DELETE FROM [SYS:KeySets] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+  exec' db "DELETE FROM [SYS:Modules] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+  exec' db "DELETE FROM [SYS:Namespaces] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+  exec' db "DELETE FROM [SYS:Pacts] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+
+{- rename VersionedTables to UserTables -}
+tableMaintenanceRowsVersionedUserTables :: BlockVersion -> BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedUserTables (BlockVersion (BlockHeight bh) _) = do
+  tblNames <- callDb $ \db ->
+    qry db "SELECT tablename FROM VersionedTables WHERE blockheight < ?;" [SInt (fromIntegral bh)] [RText]
```

Comment:
This is gonna update every table always, rather than just touching the tables that have been updated on the fork. Here's what I think you should do instead:

```SQL
CREATE TABLE VersionedTableCreation(
    tablename TEXT NOT NULL,
    blockheight BIGINT NOT NULL,
    CONSTRAINT foo UNIQUE(tablename));
CREATE TABLE VersionedTableMutation(
    tablename TEXT NOT NULL,
    blockheight BIGINT NOT NULL,
   CONSTRAINT foo2 UNIQUE(tablename, blockheight));
CREATE INDEX foo3 ON VersionedTableMutation(blockheight);
```

Every time you write a row you have to twiddle the mutation table (with a `REPLACE`). Then during this 
maintenance:

getting list of tables to `DROP`:
```
SELECT DISTINCT tablename FROM VersionedTableCreation WHERE blockheight > ?
```

getting list of tables you have to run `DELETE` on:

```
SELECT DISTINCT tablename FROM VersionedTableMutation WHERE blockheight > ?
```

Reply:
Let's hold off on this until we do on-demand, the specification for now is "pessimistic" which means no dirty tracking.